### ISSUE TITLE
Quality Assessment of RDF Datasets

### ISSUE THREAD

**Issue by Invincibled** [SANSA-Stack/SANSA-Stack#163]
![image-20210622190827391](https://user-images.githubusercontent.com/736767/123023978-f29bf280-d40a-11eb-8636-51c1898e66f5.png)

Resource list:
Luzzu Framework: https://github.com/Luzzu/Framework
Luzzu Metrics: https://github.com/Luzzu/Metrics

Background: We are looking to calculate various semantic based metrics on different RDF datasets.  We have previously been using Luzzu, but are now looking to migrate to using Sansa

Question 1: When comparing the overlapping metrics between Sansa and Luzzu, we have found several large discrepancies (see screen shot).  Do you know why there would such a large discrepency?  The differences in algorithms and implementations as described here: [http://www.semantic-web-journal.net/system/files/swj1757.pdf](http://www.semantic-web-journal.net/system/files/swj1757.pdf) not seem to indicate such a large discrepency would be expected.  For reference, the dataset we are using is less than 10 MB in size.

Question 2: We are interested in calculating the metrics in Luzzu that are not in Sansa.  What are your recommendations for porting the metrics to SANSA?  Also, what are your recommendations in general for adding additional metrics to SANSA?

**Comment by GezimSejdiu** [comment]
Hi @Invincibled , 

thanks a lot for reaching out and submitting this issue. It is great to see that you found SANSA Quality Assessment resource useful for such cases whenever the dataset becomes large to be computed via Luzzu. 

Our approach (see its implementation of some of the metrics [here](https://github.com/SANSA-Stack/SANSA-Stack/tree/develop/sansa-rdf/sansa-rdf-spark/src/main/scala/net/sansa_stack/rdf/spark/qualityassessment/metrics)) depend on a [configuration file](https://github.com/SANSA-Stack/SANSA-Stack/blob/develop/sansa-rdf/sansa-rdf-spark/src/main/resources/metrics.conf) that specify how such values are. Let's take an example of a short URL. If your setting on Luze was 80 -- all URI that is shorter than 80 characters are part of the equation, then on SANSA the default value (if you didn't change it was set to be 95). I think it would also be great if you can share the dataset and such a setting so that we can also compare why there are such discrepancies in the results. 

Reg. your question 2) -- It would be great if you can contribute to SANSA (see the [contribution guide](http://sansa-stack.net/contributing-to-sansa/) for more details) by adding/porting new metrics which aren't part of SANSA (yet) so that community can also benefit from them. 

Best regards, 
Gezim

**Comment by Invincibled** [comment]
Hi @GezimSejdiu 

First of all, thanks for your reply.

Question 1: I have modified the program according to the configuration information you provided, however, the calculation result is still different. I checked the source code and found that the algorithm is essentially different, so I don't know which algorithm we should use, do you have any suggestions?

Question 2: Since Luzzu provides more implementations of other metrics, and Luzzu is relatively weak in big data support, do you have any other implementations that you recommend (other than Sansa), or what are your recommended Sansa best practices?

Looking forward to your reply

**Comment by Invincibled** [comment]
If we want to extend new metrics, do you think we should modify the source code or use an interface to extend it?

Looking forward to your reply, thanks.

**Comment by Aklakan** [comment]
Judging from e.g. [this luzzu metric](https://github.com/Luzzu/Metrics/blob/master/linked-data-quality-metrics/contextual/src/main/java/io/github/luzzu/linkeddata/qualitymetrics/contextual/provenance/ExtendedProvenanceMetric.java) what you want to do is basically create an aggregation function for quads.

As for interfaces, from my perspective all you need is a bunch of static utility functions that create aggregators for the metrics you want as shown in the following:


So using Java8 streams, aggregation looks as follows:

```java
Stream<Quad> stream = ...;
Model resultModel = stream.collect(MyMetrics.extendedProvenance());
```

where MyMetrics.extendedProvenance() returns a java Collector - which is exactly a thing for parallel aggregation!


With the implicits provided by SANSA with the `import net.sansa_stack.query.spark._` we get a `javaCollect` method which can leverage a java8 collector:
```
RDD<Quad> rdd = ...;
Model resultModel = rdd.javaCollect(MyMetrics.extendedProvenance());
```
`javaCollect` is actually implemented by the method [aggregateUsingJavaCollector](https://github.com/SANSA-Stack/SANSA-Stack/blob/31ce79d95524ea9a1f80f6fe29612d71b2259b1e/sansa-query/sansa-query-spark/src/main/scala/net/sansa_stack/query/spark/ops/rdd/RddOfAnyOps.scala#L29).


So this means, conceptually, there is no difference between computing an aggregation form a plain Java8 stream or an RDD - and **if the collector and its generated objects are serializable** you can reuse the same code in both settings.

In order to help creating a **serializable collector** there is the AggBuilder utilty class. Among other things, it provides the `fromCollector` method. The arguments are just serializable versions of the [standard collector creation function](https://docs.oracle.com/javase/8/docs/api/java/util/stream/Collector.html#of-java.util.function.Supplier-java.util.function.BiConsumer-java.util.function.BinaryOperator-java.util.function.Function-java.util.stream.Collector.Characteristics...-):

```java
class AggBuilder {
    public static <T, R, A> AggFromCollector<T, R, A> fromCollector(
            SerializableSupplier<A> supplier,
            SerializableBiConsumer<A, T> accumulator,
            SerializableBinaryOperator<A> combiner,
            SerializableFunction<A, R> finisher) {
        ...
    }
}
```

Adapting this to the metric computation, we get:

```java
public static ParallelAgg<Quad, MyLuzzuMetric, Model> createAggForMyMetric() {
  return AggBuilder.fromCollector(
    () -> new MyLuzzuMetric(),
    (metric, quad) -> metric.compute(quad),
    (metric1, metric2) -> combine(metric1, metric2), /* !!! */
    metric -> metric.getResultModel());
}
```

with usage:
```java
Model result = rdd.javaCollect(createAggForMyMetric().asCollector())
```

The crucial part is marked with `/* !!! */`. If for a metric you cannot devise a method to combine two partial results then the computation won't benefit from parallelization.

The main power of the AggBuilder is that it features methods to compose aggragators.
For example, let's say you want to have a map with multiple aggregators that should run in a single pass, then you can do:

```java
Map<String, ParallelAggregator<Quad, Model, ?>> metrics = new HashMap<>();
metrics.put("extendedProvenance", MyMetrics.extendedProvenance());
metrics.put("someOtherMetricWithModelResult", MyMetrics.foobar());

// inputBroadcastMap sends each item to every aggregator in a map where the values are aggregators.
Map<String, Model> result = rdd.javaCollect(AggBuilder.inputBroadcastMap(metrics).asCollector())
```

In a nutshell, the approach using the AggBuilder provides these benefits:

* Can be used both with plain java8 stream parallelization (without spark) and with spark - makes implementations more reusable.
* Can assemble computation of multiple metrics into a single task using `inputBroadcastMap`. This lack of composition is a limitation of the current design of the metrics in Sansa, yet there are already some parts in Sansa that rely on this newer system.

I hope that helps for your considerations.

**Comment by Aklakan** [comment]
What we actually need is some kind of execution graph model which metrics can be implemented against such that computation results (e.g. number of triples) could be shared among metrics. @LorenzBuehmann you once mentioned that something like this was started in the inference module. Would it be feasible to extend on that or would we need a different existing or custom solution?
